### ISSUE TITLE
Problems with Tensor type and transforms

### ISSUE THREAD

**Issue by JonasHarrerFLoy** [fepegar/torchio#822]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem summary

Similar problems already occured in: https://github.com/fepegar/torchio/issues/647

I have two problems when trying to create a minimal working dataloader for my data.
1. If I try to load the data just using tio.ToCanonical() and  tio.EnsureShapeMultiple(16), is there another transform like torchvision toTensor(). I always get the error that the image is of the wrong type (int Tensor instead of float Tensor). Do I need to cast float manually?
2. If I add tio.ZNormalization() or tio.RescaleIntensity() problem 1. is fixed. But I then get the problem with the inconsistent space. Eventhough I am using ToCanonical in the first place. Can this occur due to the wrong order of transforms?

I appreciate any help on this! Thank you! :)

### Code for reproduction

```python
from monai.networks.nets import UNet
import torchio as tio
from torch.utils.data import DataLoader


# problem 1
trafo = tio.Compose([tio.ToCanonical(), tio.EnsureShapeMultiple(16)])

# or for problem 2
trafo = tio.Compose([tio.ToCanonical(), tio.ZNormalization(), tio.EnsureShapeMultiple(16)])

# this is stolen from issue 647:
class MySubject(tio.Subject):
    def check_consistent_attribute(self, *args, **kwargs) -> None:
        kwargs['relative_tolerance'] = 1e-5
        kwargs['absolute_tolerance'] = 1e-5
        return super().check_consistent_attribute(*args, **kwargs)

# Iterate through all nifti paths
data = []    
for sample_type in samples:
        for path in samples[sample_type]:
            data.append(MySubject(
                image=tio.ScalarImage(os.path.join(path)),
                label=tio.LabelMap(os.path.join(samples[sample_type][path]['mask']))
            ))
            
# create SubjectDataset and dataloader
dataset = tio.SubjectsDataset(data, transform=trafo)
dataload = DataLoader(dataset, batch_size=1)

# UNet from Monai:
net = UNet(
            spatial_dims=3,
            in_channels=1,
            out_channels=1,
            channels=(16, 32, 64, 128, 256),
            strides=(2, 2, 2, 2),
            num_res_units=2,
            norm=Norm.BATCH,
        )

for data in dataload:
    net(data['image'][tio.DATA])
```

### Actual outcome

Code doesn't run. See error messages below

### Error messages

```python-traceback
Error message for 1: _RuntimeError: Input type (torch.cuda.DoubleTensor) and weight type (torch.cuda.FloatTensor) should be the same_
Error message for 2.: _As described above, some images in the subject are not in the same space. You probably can use the transforms ToCanonical and Resample to fix this, as explained at https://github.com/fepegar/torchio/issues/647#issuecomment-913025695_
```

### Expected outcome

Code running and network outputting the correct format

### System info

```shell
# flake8: noqa

import re
import sys
import platform
import torchio
import torch
import numpy
import SimpleITK as sitk


sitk_version = re.findall('SimpleITK Version: (.*?)\n', str(sitk.Version()))[0]

print('Platform:  ', platform.platform())
print('TorchIO:   ', torchio.__version__)
print('PyTorch:   ', torch.__version__)
print('SimpleITK: ', sitk_version)
print('NumPy:     ', numpy.__version__)
print('Python:    ', sys.version)
```

**Comment by romainVala** [comment]
Hi
thanks for sharing

this is indeed weird to have some torch type issue ...
@fepegar is the tensor type explicitly set for the nifti data type ?
and which transform does modify the tesor type  (I would expect that `ZNormalization` will force the output type to be float) 
but why is there an issue with torch.cuda.DoubleTensor and torch.cuda.FloatTensor ?

if you can share a minimal data exemple so that I could reproduce the issue ? if not possible may be you can reproduce it from the torchio colin dataset by adding so perturbation to the affine and changing the input datatype ...

**Comment by JonasHarrerFLoy** [comment]
Hi,
thank you for the fast answer!

Sadly I am not allowed to share my data. I will see if it also occurs with other nifti MRI data (colin dataset) or if this is only happening with my data.
I am using T1 Sagittal MRI series.
As you mentioned the type problem is not showing up when doing the normalization. With normalization the type is correct, but the Error message 2 comes up (wrong space, toCanonical).

**Comment by romainVala** [comment]
Actually I only need the header information (not the voxel value) so to share it may be you can
load the data, multiply alll by 0 and save it again (check that the type and affine in the nifti is still the same as the original) and that you reproduce the bug with it ...

**Comment by JonasHarrerFLoy** [comment]
Thank you!

I prepared a file to test it: 

[test_nifti.nii.gz](https://github.com/fepegar/torchio/files/8088241/test_nifti.nii.gz)

**Comment by romainVala** [comment]
you just send one volume, to reproduce I need 2 (an image and the label, with a different affine ) ... non ?

**Comment by JonasHarrerFLoy** [comment]
I am sorry, here is the corresponding mask:
[test_nifti-mask.nii.gz](https://github.com/fepegar/torchio/files/8088849/test_nifti-mask.nii.gz)

**Comment by JonasHarrerFLoy** [comment]
Hey romainVala,

thank you so much! 
It works :)

Have a nice weekend.
